Make goto_rw less conservative

karkhaz opened the pull request:
When handling the case of pointers, goto_rw would mark both the object pointed
to and the pointer itself as written. For example, if x = &i, the line `*x = 1'
would yield a write set containing both x and i, instead of just i.

This resolves that issue by ensuring that value_set_dereference always gives at
least one object that the dereference can refer to.

This also resolves a bug wherein &a is marked as reading from a by removing the
ID_symbol special case from get_objects_address_of.

karkhaz commented:
(CI passing is currently blocked on  #4922)

A couple of questions, not sure who the right person to answer this is:

this looks like a performance improvement, not something that would actually affect the verification result---right? In which case I'm not sure if I should write a test, or how I would do so.

There is a TODO in the code, not sure whether to leave it in or what I might do about it.

thank you!

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 872cc8c486243fff82fd724c32b6a6026fb3c086).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/119522812

smowton commented:
Let's have some regression tests illustrating the change? This is no worry for symex, as it doesn't use dynamic objects this way (it makes up symbols for them instead)

tautschnig commented on src/analyses/reaching_definitions.cpp:
```diff
@@ -321,20 +321,17 @@ void rd_range_domaint::transform_assign(
-         !rd.get_is_dirty()(identifier))))
+    if(
+      is_must_alias && (!rd.get_is_threaded()(from) ||
+                        (symbol_ptr != nullptr && symbol_ptr->is_shared() &&
```
Maybe we should do `(!not_found && symbol_ptr->is_shared() && ...`

tautschnig commented on src/pointer-analysis/value_set_dereference.cpp:
```diff
@@ -398,8 +398,12 @@ value_set_dereferencet::valuet value_set_dereferencet::build_reference_to(
+    // successfully ensures that every (non-null) dereference points to
+    // something, making the write set computation work correctly, but dynamic
+    // objects do not seem to be intended to be used in this way.
+    result.value = object;
```
The problem is the offset, which might not be known. I believe this should be correct when `o.offset().is_constant() == true`. I think in all other cases you largely need to fall back to what the `byte_extract` base does below.

karkhaz committed:
Make writes to pointees independent of pointer

The write set of a pointer should not include writes to the pointed-to
object. This commit makes that true.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 29f8f6986041d0b61e670dafb74dfc3be2c86308).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122637868
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

karkhaz commented:
@smowton thanks. Some of the patch was causing a regression, so I've included the most relevant part and added a test case. Thanks to @tautschnig for help with the test.

karkhaz commented:
@peterschrammel may I have a review of this please? Although I won't be able to merge until @tautschnig gets back and approves...

karkhaz commented:
@tautschnig are you happy with the changes that I've made?
